ledger/acctupdates_test.go

```diff
@@ -1637,3 +1637,256 @@ func TestCachesInitialization(t *testing.T) {
+
+	au := &accountUpdates{}
+	au.initialize(config.GetDefaultLocal(), ".", initialProtoParams, accts[0])
+	err := au.loadFromDisk(ml)
+	require.NoError(t, err)
+	defer au.close()
+
+	// cover initialRounds genesis blocks
+	rewardLevel := uint64(0)
+	for i := 1; i < int(initialRounds); i++ {
+		accts = append(accts, accts[0])
+		rewardsLevels = append(rewardsLevels, rewardLevel)
+	}
+
+	extraRounds := uint64(39)
+
+	// write extraRounds rounds so that we will fill up the queue.
+	for i := basics.Round(initialRounds); i < basics.Round(initialRounds+extraRounds); i++ {
+		rewardLevelDelta := crypto.RandUint64() % 5
+		rewardLevel += rewardLevelDelta
+		accountChanges := 2
+
+		updates, totals := randomDeltasBalanced(accountChanges, accts[i-1], rewardLevel)
+		prevTotals, err := au.Totals(basics.Round(i - 1))
+		require.NoError(t, err)
+
+		newPool := totals[testPoolAddr]
+		newPool.MicroAlgos.Raw -= prevTotals.RewardUnits() * rewardLevelDelta
+		updates.Upsert(testPoolAddr, newPool)
+		totals[testPoolAddr] = newPool
+
+		blk := bookkeeping.Block{
+			BlockHeader: bookkeeping.BlockHeader{
+				Round: basics.Round(i),
+			},
+		}
+		blk.RewardsLevel = rewardLevel
+		blk.CurrentProtocol = initProtocolVersion
+
+		delta := ledgercore.MakeStateDelta(&blk.BlockHeader, 0, updates.Len(), 0)
+		delta.Accts.MergeAccounts(updates)
+		ml.addMockBlock(blockEntry{block: blk}, delta)
+		au.newBlock(blk, delta)
+		accts = append(accts, totals)
+		rewardsLevels = append(rewardsLevels, rewardLevel)
+	}
+
+	newVersion := protocol.ConsensusV21
+	// add MaxBalLookback-extraRounds more rounds that contains blocks using a newer consensus version.
+	for i := basics.Round(initialRounds + extraRounds); i <= basics.Round(initialRounds+extraRounds+initialProtoParams.MaxBalLookback); i++ {
+		rewardLevelDelta := crypto.RandUint64() % 5
+		rewardLevel += rewardLevelDelta
+		accountChanges := 2
+
+		updates, totals := randomDeltasBalanced(accountChanges, accts[i-1], rewardLevel)
+		prevTotals, err := au.Totals(basics.Round(i - 1))
+		require.NoError(t, err)
+
+		newPool := totals[testPoolAddr]
+		newPool.MicroAlgos.Raw -= prevTotals.RewardUnits() * rewardLevelDelta
+		updates.Upsert(testPoolAddr, newPool)
+		totals[testPoolAddr] = newPool
+
+		blk := bookkeeping.Block{
+			BlockHeader: bookkeeping.BlockHeader{
+				Round: basics.Round(i),
+			},
+		}
+		blk.RewardsLevel = rewardLevel
+		blk.CurrentProtocol = newVersion
+
+		delta := ledgercore.MakeStateDelta(&blk.BlockHeader, 0, updates.Len(), 0)
+		delta.Accts.MergeAccounts(updates)
+		ml.addMockBlock(blockEntry{block: blk}, delta)
+		au.newBlock(blk, delta)
+		accts = append(accts, totals)
+		rewardsLevels = append(rewardsLevels, rewardLevel)
+	}
+	// now, commit and verify that the committedUpTo method broken the range correctly.
+	au.committedUpTo(basics.Round(initialRounds + extraRounds + initialProtoParams.MaxBalLookback))
+	au.waitAccountsWriting()
+	require.Equal(t, basics.Round(initialRounds+extraRounds)-1, au.dbRound)
```

Comment:
Who is this loop different from the last one from the prev test? there is an additional `newVersionBlocksCount` value but it looks not really needed since we add more than `initialRounds+extraRounds)-1` and check that only `initialRounds+extraRounds)-1` is committed at single `committedUpTo` call.

Reply:
This test ensures that we know not only to break the rounds at the right place, but also to continue from the right place on the subsequent write.

Now that the code is properly written, it would be impossible for you to create a case where it's not happening, but during development it did, and this was the test I used to verify it against.

Reply:
btw - I agree that there is no need for writing `extraRounds` here. One round would be enough. But for verification purposes, having multiple is usually better than having just one.